src/runtime/slots/mp_length.cs

```diff
@@ -0,0 +1,66 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace Python.Runtime.Slots
+{
+    internal static class mp_length_slot
+    {
+        /// <summary>
+        /// Finds a get_Count (Count property) that is private and
+        /// is either ICollection or ICollection&lt;T&gt; explicit
+        /// implementation
+        /// </summary>
+        static MethodInfo GetCountGetter(Type t)
+        {
+            foreach (var info in t.GetMethods(BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.DeclaredOnly))
+            {
+                if (info.IsFinal && info.IsPrivate && (info.Name.Contains("System.Collections.Generic.ICollection") || info.Name.Contains("System.Collections.ICollection")) && info.Name.Contains("get_Count"))
+                {
+                    return info;
+                }
+            }
+            return null;
+        }
+
+        /// <summary>
+        /// Implements __len__ for classes that implement ICollection
+        /// (this includes any IList implementer or Array subclass)
+        /// </summary>
+        public static int mp_length(IntPtr ob)
+        {
+            var co = ManagedType.GetManagedObject(ob) as CLRObject;
+            if (co == null)
+            {
+                Exceptions.RaiseTypeError("invalid object");
+            }
+
+            // first look for ICollection implementation directly
+            if (co.inst is ICollection c)
+            {
+                return c.Count;
+            }
+
+            Type clrType = co.inst.GetType();
+
+            // now look for things that implement ICollection<T> directly (non-explicitly)
+            PropertyInfo p = clrType.GetProperty("Count");
+            if (p != null && clrType.GetInterfaces().Any(x => x.IsGenericType && x.GetGenericTypeDefinition() == typeof(ICollection<>)))
```

Comment:
Too many reflections every times, you can try like this.
```c#
    class Program
    {
        interface ICountGetter
        {
            int GetCount(object obj);
        }

        class CountGetter<Cls, T> : ICountGetter
        {
            private Func<Cls, int> _getter;

            public CountGetter(Func<Cls, int> getter)
            {
                _getter = getter;
            }

            public int GetCount(object obj)
            {
                return _getter((Cls)obj);
            }
        }

        static void Main(string[] args)
        {
            var obj = new MyClass();
            var type = obj.GetType();
            var declType = type.GetInterfaces().FirstOrDefault(
                t => t.IsGenericType
                && t.GetGenericTypeDefinition() == typeof(ICollection<>));
            if (declType == null)
            {
                // set error and return
            }
            var pi = declType.GetProperty("Count");
            var elemType = declType.GenericTypeArguments[0];
            var getterType = typeof(CountGetter<,>).MakeGenericType(declType, elemType);
            var getterFuncType = typeof(Func<,>).MakeGenericType(declType, typeof(int));
            var getterFunc = Delegate.CreateDelegate(getterFuncType, pi.GetMethod);
            var getter = (ICountGetter)Activator.CreateInstance(getterType, getterFunc);
            // You should cache the ([decltype] = getter), it makes you no reflection costs when you call GetCount next time.
            int count = getter.GetCount(obj);
            Console.WriteLine(count);
        }
    }
```

Reply:
I agree, but I feel like it would be good to have a feature + tests first. It can be optimized later.